ecosystem/indexer-grpc/indexer-grpc-table-info/src/table_info_parser.rs

```diff
@@ -0,0 +1,217 @@
+                    max_version = std::cmp::max(max_version, version);
+                },
+                Err(err) => {
+                    return Err(err);
+                },
+            }
+        }
+        Ok(max_version)
+    }
+
+    async fn get_batches(&mut self) -> Vec<TransactionBatchInfo> {
+        self.ensure_highest_known_version().await;
+
+        info!(
+            current_version = self.current_version,
+            highest_known_version = self.highest_known_version,
+            parser_batch_size = self.parser_batch_size,
+            parser_task_count = self.parser_task_count,
+            "Preparing to fetch transactions"
+        );
+
+        let mut starting_version = self.current_version;
+        let mut num_fetches = 0;
+        let mut batches = vec![];
+
+        while num_fetches < self.parser_task_count && starting_version <= self.highest_known_version
+        {
+            let num_transactions_to_fetch = std::cmp::min(
+                self.parser_batch_size as u64,
+                self.highest_known_version - starting_version + 1,
+            ) as u16;
+
+            batches.push(TransactionBatchInfo {
+                start_version: starting_version,
+                num_transactions_to_fetch,
+            });
+            starting_version += num_transactions_to_fetch as u64;
+            num_fetches += 1;
+        }
+        batches
+    }
+
+    async fn fetch_raw_txns_with_retries(
+        context: Arc<Context>,
+        ledger_version: u64,
+        batch: TransactionBatchInfo,
+    ) -> Vec<TransactionOnChainData> {
+        loop {
+            match context.get_transactions(
+                batch.start_version,
+                batch.num_transactions_to_fetch,
+                ledger_version,
+            ) {
+                Ok(raw_txns) => return raw_txns,
+                Err(_err) => {
+                    continue;
+                },
+            }
+        }
+    }
+
+    /// Parses the table information from the raw transactions before converting to the api transactions,
+    /// optionally backup the rocksdb to gcs depending on epoch advancement or not.
+    async fn parse_table_info(
+        context: Arc<Context>,
+        raw_txns: Vec<TransactionOnChainData>,
+        db_writer: Arc<dyn DbWriter>,
+    ) -> Result<(), Error> {
+        if raw_txns.is_empty() {
+            return Ok(());
+        }
+
+        let start_millis = chrono::Utc::now().naive_utc();
+        let first_version = raw_txns.first().map(|txn| txn.version).unwrap();
+        let ledger_chain_id = context.chain_id().id();
+        let (_, _, block_event) = context
+            .db
+            .get_block_info_by_version(first_version)
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Could not get block_info for start version {}",
+                    first_version,
+                )
+            });
+        let block_event_epoch = block_event.epoch();
```

Comment:
Why do you need this?

Reply:
in the next PR in backup and restore, we use epoch instead of versions for rocksdb snapshot upload, but i assume i could get rid of it in this PR. context: use epoch over txn cuz for state syncing, it syncs one epoch before, so making it epoch based.